app/services/subdomain_extractor.rb

```
@@ -0,0 +1,19 @@
+class SubdomainExtractor
+    def self.extract(request)
+      subdomain = request.subdomain.presence || 
+                  request.params[:subdomain] ||
+                  extract_from_host(request.host)
+      Rails.logger.debug "Extracted subdomain: #{subdomain}"
+      subdomain
+    end
+  
+    private
+  
+    def self.extract_from_host(host)
```

RuboCop (1.69.1)

[warning] `private` (on line 11) does not make singleton methods private. Use `private_class_method` or `private` inside a `class << self` block instead.

(Lint/IneffectiveAccessModifier)